### ISSUE TITLE
needless_match can trigger from some single-arm capture matches

### ISSUE THREAD

**Issue by Fishrock123** [rust-lang/rust-clippy#11236]
### Summary

The `needless_match` lint can apparently trigger incorrectly in some edge cases involving capturing matches.

This could be a wild goose chase on what the root cause could be so I'd rather not chase it down much, someone with a better idea would be better to take a look.

I think the gist is that is you capture the whole match but only on a discriminant, the lint might think you are just always trying to return the value, which is not correct.

### Lint Name

needless_match

### Reproducer

I tried this code:

```rust
let ord = match self
    .x
    .partial_cmp(&other.string_match_param)
{
    ord @ Some(Ordering::Equal) => ord,
    other => return other,
};
```

I saw this happen:

```
warning: this match expression is unnecessary
   --> src/x.rs:565:19
    |
565 |           let ord = match self
    |  ___________________^
566 | |             .string_match_param
567 | |             .partial_cmp(&other.x)
568 | |         {
569 | |             ord @ Some(Ordering::Equal) => ord,
570 | |             other => return other,
571 | |         };
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_match
    = note: `#[warn(clippy::needless_match)]` on by default
help: replace it with
    |
565 ~         let ord = self
566 +             .x
567 ~             .partial_cmp(&other.x);
    |

```

I expected to see this happen:

No Error - this match serves are purpose and cannot be eliminated like that.

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: aarch64-apple-darwin
release: 1.70.0
LLVM version: 16.0.2
```
